store.rb

```diff
@@ -118,4 +153,26 @@ def get_node(node)
       raise e
     end
   end
+
+  # immediately update cache if node joins/leaves
+  def watch()
+    return if @zk.nil?
+    @root_watcher ||= @zk.register("/", :only => :child) do |event|
+      @log.info "Children added/deleted"
+      reload_instances
+    end
+  end
+
+  def reload_instances()
+    instances = load_instances_from_zk()
+    cache_instances(instances)
+  end
+
+  def get_cached_instances()
+    @mutex.synchronize { @cached_nodes }
+  end
+
+  def cache_instances(instances)
+    @mutex.synchronize { @cached_nodes = instances }
```

Comment:
Technically, I think these mutexes are unnecessary as there isn't an interleaving that leads to an inconsistency.  MRI Ruby would also wrap this execution in the GVL. I think these also fail to ensure an ordering where the freshest data wins, because the lock is only around the assignment of the variable.

I think a write lock only that wraps around the fetching from ZK as well as the assignment would probably provide the guarantee that we want.  I'm not super sure about any of this, so I'd love @nelgau's opinion here.